### PR TITLE
New API for top-level updates

### PR DESCRIPTION
**Pull request by acdlite**
Introduces an additional, new API for performing top-level updates in React DOM:

```js
// Existing API
ReactDOM.render(<App foo="A" />, container); // mount
ReactDOM.render(<App foo="B" />, container); // update
ReactDOM.unmountComponentAtNode(container); // unmount


// New API
const root = ReactDOM.unstable_createRoot(container);

// Simple
root.render(<App foo="A" />); // mount
root.render(<App foo="B" />); // update
root.unmount(); // unmount

// With pre-rendering
const work = root.prerender(<App foo="C" />); // Begin rendering but don't flush yet.
await work; // Wait for work to complete. Still don't flush yet.
work.commit(); // Synchronously flush changes

// Start prerendering before container is available
let container;
const root = ReactDOM.unstable_createRoot(function getContainer() {
  return container;
});
root.prerender(<App foo="A" />); // Begin rendering but don't flush yet.
container = await promiseForContainer;
work.commit(); // Synchronously flush remaining work
```

(Will fill in with more details later regarding motivations and use cases.)

I had thought we might be able to implement the existing API (`ReactDOM.render`) on top of the new API, but the semantics of the `.then` callback are sufficiently different that I don't think it's possible without a breaking change.

Work-in-progress:

- [x] `ReactDOM.unstable_createRoot` that returns a public root instance. Replaces hidden `_reactRootContainer` field.
- [x] Work type that represents a top-level update. Has `then` and `commit` methods.
- [x] `work.commit()` synchronously flushes all remaining work.
- [x] Pre-rendering in one tree should not block updates in a separate tree.
- [x] Completion callbacks (`.then`) resolve synchronously if tree is already complete
- [x] Committing one tree should not flush updates in a separate tree.
- [x] Figure out how hydration fits into this.
- [x] Lazy containers
  - [x] Namespace parameter
  - [x] Owner document parameter
- [x] Things that are "done" but need tests to confirm
  - [x] Rendering before DOM container is available

Depends on expiration times PR https://github.com/facebook/react/pull/10426

### CODE REVIEW

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js [hunk 1]
```diff
@@ -646,7 +691,70 @@ function renderSubtreeIntoContainer(
...
+  work.then(() => {
+    work.commit();
+    if (callback !== null) {
+      (callback: any)();
+    }
+  });
+};
+
 var ReactDOMFiber = {
+  unstable_create(
```
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: As much as I like the `flushSync(batch)` naming convention, this might be a bit too overloaded and vague.

`unstable_createRoot`?

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js [hunk 2]
```diff
@@ -646,7 +691,70 @@ function renderSubtreeIntoContainer(
...
+    container: DOMContainer | (() => DOMContainer),
+    namespace: ?string,
```
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: Isn't this suppose to be `namespace ?: string`?

File: fixtures/fiber-triangle/index.html [hunk 3]
```diff
@@ -76,51 +76,54 @@
...
-      function SierpinskiTriangle({ x, y, s, children }) {
-        if (s <= targetSize) {
-          return (
-            <Dot
-              x={x - (targetSize / 2)}
-              y={y - (targetSize / 2)}
-              size={targetSize}
-              text={children}
-            />
+      class SierpinskiTriangle extends React.Component {
```
Comment (sebmarkbage) [inline fixtures/fiber-triangle/index.html]: Unrelated?

**Comment by sebmarkbage** [comment]
Did we figure out how the `hydrate` API fits into this?

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js (hunk 2 above)
Comment (acdlite) [reply src/renderers/dom/fiber/ReactDOMFiberEntry.js]: Ooh, didn't know Flow treated those differently.

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js (hunk 1 above)
Comment (acdlite) [reply src/renderers/dom/fiber/ReactDOMFiberEntry.js]: Works for me. (I figure we can have a team bikeshedding meeting before we land this :D)

**Comment by acdlite** [comment]
Nope, I'll add that to the todos

File: src/renderers/shared/fiber/__tests__/ReactIncrementalRoot-test.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  it('prerenders roots', () => {
+    const root = ReactNoop.create();
+    const work = root.prerender(<span prop="A" />);
+    expect(root.getChildren()).toEqual([]);
+    work.commit();
+    expect(root.getChildren()).toEqual([span('A')]);
+  });
+
+  it('resolves `then` callback synchronously if tree is already completed', () => {
```
Comment (acdlite) [inline src/renderers/shared/fiber/__tests__/ReactIncrementalRoot-test.js]: Hmm, is this behavior really necessary for async updates? Or is it only relevant to sync?

**Comment by gaearon** [comment]
Is it not confusing that something that looks thenable doesn't behave like a promise (zalgo)?

**Comment by acdlite** [comment]
@gaearon We thought it was valuable to be able to implement the normal `root.render` API on top of the `root.prerender` API, including in sync mode. It might be confusing, but I think we can reasonably assume that `prerender` is an advanced, lower-level API that will be used be relatively few people, at least directly. Also, if you await a thenable, it's automatically wrapped in a promise.

**Comment by acdlite** [comment]
@sebmarkbage Starts with commit titled "Initial version of blocking"

**Comment by gaearon** [comment]
Seems stale and superseded by https://github.com/facebook/react/pull/11225.

**Comment by NE-SmallTown** [comment]
@acdlite Is there any explanation why we drop this?
